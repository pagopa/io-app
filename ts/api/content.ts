/**
 * This module implements the types and a client for retrieving the static
 * content published at https://github.com/teamdigitale/italia-services-metadata
 */

import {
  basicResponseDecoder,
  BasicResponseType,
  createFetchRequestForApi,
  IGetApiRequestType
} from "italia-ts-commons/lib/requests";

import { contentRepoUrl } from "../config";

import { defaultRetryingFetch } from "../utils/fetch";

import { ServiceId } from "../../definitions/backend/ServiceId";
import { Service as ServiceMetadata } from "../../definitions/content/Service";

type GetServiceT = IGetApiRequestType<
  {
    serviceId: ServiceId;
  },
  never,
  never,
  BasicResponseType<ServiceMetadata>
>;

const getServiceT: GetServiceT = {
  method: "get",
  url: params => `/services/${params.serviceId}.json`,
  query: _ => ({}),
  headers: _ => ({}),
  response_decoder: basicResponseDecoder(ServiceMetadata)
};

/**
 * A client for the static content
 */
export function ContentClient(fetchApi: typeof fetch = defaultRetryingFetch()) {
  const options = {
    baseUrl: contentRepoUrl,
    fetchApi
  };

  return {
    getService: createFetchRequestForApi(getServiceT, options)
  };
}
