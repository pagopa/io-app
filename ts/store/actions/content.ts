import {
  CONTENT_SERVICE_LOAD,
  CONTENT_SERVICE_LOAD_FAILURE,
  CONTENT_SERVICE_LOAD_SUCCESS
} from "./constants";

import { Service as ServiceMetadata } from "../../../definitions/content/Service";

import { ServiceId } from "../../../definitions/backend/ServiceId";

export type ContentServiceLoad = Readonly<{
  type: typeof CONTENT_SERVICE_LOAD;
  serviceId: ServiceId;
}>;

type ContentServiceLoadSuccess = Readonly<{
  type: typeof CONTENT_SERVICE_LOAD_SUCCESS;
  serviceId: ServiceId;
  data: ServiceMetadata;
}>;

type ContentServiceLoadFailure = Readonly<{
  type: typeof CONTENT_SERVICE_LOAD_FAILURE;
  serviceId: ServiceId;
}>;

export type ContentActions =
  | ContentServiceLoad
  | ContentServiceLoadSuccess
  | ContentServiceLoadFailure;

export const contentServiceLoad = (
  serviceId: ServiceId
): ContentServiceLoad => ({
  type: "CONTENT_SERVICE_LOAD",
  serviceId
});

export const contentServiceLoadSuccess = (
  serviceId: ServiceId,
  data: ServiceMetadata
): ContentServiceLoadSuccess => ({
  type: "CONTENT_SERVICE_LOAD_SUCCESS",
  serviceId,
  data
});

export const contentServiceLoadFailure = (
  serviceId: ServiceId
): ContentServiceLoadFailure => ({
  type: "CONTENT_SERVICE_LOAD_FAILURE",
  serviceId
});
