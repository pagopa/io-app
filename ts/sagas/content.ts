import { call, Effect, fork, put, take } from "redux-saga/effects";

import { BasicResponseType } from "italia-ts-commons/lib/requests";

import { ContentClient, ServiceMetadata } from "../api/content";
import {
  ContentServiceLoad,
  contentServiceLoadFailure,
  contentServiceLoadSuccess
} from "../store/actions/content";

import { CONTENT_SERVICE_LOAD } from "../store/actions/constants";

export function* contentSaga(): IterableIterator<Effect> {
  const contentClient = ContentClient();

  function getService(
    serviceId: string
  ): Promise<BasicResponseType<ServiceMetadata> | undefined> {
    return new Promise((resolve, _) =>
      contentClient
        .getService({ serviceId })
        .then(resolve, () => resolve(undefined))
    );
  }

  while (true) {
    const loadAction: ContentServiceLoad = yield take(CONTENT_SERVICE_LOAD);

    const serviceId = loadAction.serviceId;

    const response: BasicResponseType<ServiceMetadata> | undefined = yield call(
      getService,
      serviceId
    );

    if (response && response.status === 200) {
      yield put(contentServiceLoadSuccess(serviceId, response.value));
    } else {
      yield put(contentServiceLoadFailure(serviceId));
    }
  }
}
