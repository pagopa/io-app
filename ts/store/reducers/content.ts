/**
 * Implements the reducers for static content.
 *
 * TODO: add eviction of old entries
 * https://www.pivotaltracker.com/story/show/159440294
 */

import { CONTENT_SERVICE_LOAD_SUCCESS } from "../actions/constants";
import { Action } from "../actions/types";

import { Service as ServiceMetadata } from "../../../definitions/content/Service";

/**
 * Stores useful content such as services and organizations metadata,
 * help pages, etc...
 */
export type ContentState = Readonly<{
  servicesMetadata: {
    byId: {
      [key: string]: ServiceMetadata | undefined;
    };
  };
}>;

const initialContentState: ContentState = {
  servicesMetadata: {
    byId: {}
  }
};

export default function content(
  state: ContentState = initialContentState,
  action: Action
): ContentState {
  if (action.type === CONTENT_SERVICE_LOAD_SUCCESS) {
    return {
      ...state,
      servicesMetadata: {
        byId: {
          ...state.servicesMetadata.byId,
          [action.serviceId]: action.data
        }
      }
    };
  }
  return state;
}
