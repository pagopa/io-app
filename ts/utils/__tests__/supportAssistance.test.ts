import {
  anonymousAssistanceAddressWithSubject,
  canShowHelp,
  handleSendAssistanceLog
} from "../supportAssistance";
import { ToolEnum } from "../../../definitions/content/AssistanceToolConfig";
import MockZendesk from "../../__mocks__/io-react-native-zendesk";

jest.mock("../../config", () => ({ zendeskEnabled: true }));
describe("anonymousAssistanceAddress", () => {
  it("if the subcategory is undefined should return the email link with only the category as subject", () => {
    const mockedEmailLink = "mailto:io@assistenza.pagopa.it?subject=";
    const mockedCategory = "mockedCategory";

    expect(anonymousAssistanceAddressWithSubject(mockedCategory)).toBe(
      mockedEmailLink + mockedCategory
    );
  });
  it("if the subcategory is defined should return the email link with the category and the subcategory as subject", () => {
    const mockedEmailLink = "mailto:io@assistenza.pagopa.it?subject=";
    const mockedCategory = "mockedCategory";
    const mockedSubcategory = "mockedSubcategory";

    expect(
      anonymousAssistanceAddressWithSubject(mockedCategory, mockedSubcategory)
    ).toBe(`${mockedEmailLink}${mockedCategory}: ${mockedSubcategory}`);
  });
});

describe("canShowHelp", () => {
  it("if assistanceTool is Zendesk, should return true if the email is validated", () => {
    expect(canShowHelp(ToolEnum.zendesk, true)).toBeTruthy();
    expect(canShowHelp(ToolEnum.zendesk, false)).toBeFalsy();
  });
  it("if assistanceTool is instabug, web or none, should return false", () => {
    expect(canShowHelp(ToolEnum.instabug, true)).toBeFalsy();
    expect(canShowHelp(ToolEnum.instabug, false)).toBeFalsy();
    expect(canShowHelp(ToolEnum.web, true)).toBeFalsy();
    expect(canShowHelp(ToolEnum.none, true)).toBeFalsy();
    expect(canShowHelp(ToolEnum.web, false)).toBeFalsy();
    expect(canShowHelp(ToolEnum.none, false)).toBeFalsy();
  });
});

describe("handleSendAssistanceLog", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  it("if the assistanceTool is Zendesk should call the appendLog function", () => {
    handleSendAssistanceLog(ToolEnum.zendesk, "mockedLog");
    expect(MockZendesk.appendLog).toBeCalled();
  });
});
