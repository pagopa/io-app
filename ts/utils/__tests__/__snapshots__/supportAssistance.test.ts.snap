// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots zendeskSendCategory 1`] = `
{
  "description": {
    "de-DE": "SEND",
    "en-EN": "SEND",
    "it-IT": "SEND",
  },
  "value": "io_send",
}
`;
